<launch>

    <arg name="lidar_frame" default="/velodyne" />
    <arg name="lidar_topic" default="/velodyne_points" />

    <arg name="ground_topic" default="/ray_filter/velodyne_points_ground"/>
    <arg name="no_ground_topic" default="/ray_filter/velodyne_points_costmap"/>
    <arg name="no_ground_full_height_topic" default="/ray_filter/velodyne_points_filtered"/>


    <!-- we use 16 lidar -->
    <arg name="sensor_model" default="16"/>
    <!-- the height of lidar position  -->
    <arg name="sensor_height" default="1.70"/>
    <!-- 激光角度微分值 : degree 0.18-->
    <arg name="radial_divider_angle" default="0.18"/>
    <!-- 距离（半径）微分值 0.01-->
    <arg name="concentric_divider_distance" default="0.02"/>

    <!-- 两点之间的最大坡度 degree : degree-->
    <arg name="local_max_slope" default="8.0"/>
    <!-- 一条射线上所有点的最大坡度 : degree -->
    <arg name="general_max_slope" default="7.0"/>
    <!-- 最小高度阈值minimum height threshold -->
    <arg name="min_height_threshold" default="0.05"/>
    <!--  check if previous point is too far from previous one, if so classify again -->
    <arg name="reclass_distance_threshold" default="0.2"/>

    <!-- 高度截取: 指过滤掉高于激光雷达clip_height高度的点 -->
    <arg name="clip_height" default="0.5"/>
    <!-- 按照半径过滤点云:意指过滤掉距离小于min_distance的点，因为这些点可能是车身的反射点 -->
    <arg name="min_distance" default="0.5"/>
    <!-- XY方向点云截取（if矩形） -->
    <arg name="minX" default="-5.0"/>
    <arg name="maxX" default="20.0"/>
    <arg name="minY" default="-10.0"/>
    <arg name="maxY" default="10.0"/>

    <!-- 用于和SENSOR_HEIGHT一起，过滤近距离处的地面点 -->
    <!-- 即假设障碍物高于设定值 -->
    <!-- <arg name="default_cluster_height" default="0.05"/>  
    <arg name="default_plane_radius" default="0.1"/> -->

    <!-- DEBUG: 同时发布地面和非地面点云，以不同颜色区分 -->
    <arg name="debug_pub_all" default="true"/>
    <arg name="debug_map_topic" default="/debug/ray_filter_points"/>

    <node pkg="ray_ground_filter" type="ray_ground_filter_node" name="ray_ground_filter_node" output="screen">
        <param name="lidar_frame" value="$(arg lidar_frame)"/>
        <param name="lidar_topic" value="$(arg lidar_topic)"/>
        <param name="ground_topic" value="$(arg ground_topic)"/>
        <param name="no_ground_topic" value="$(arg no_ground_topic)"/>
        <param name="no_ground_topic_full" value="$(arg no_ground_full_height_topic)"/>
        <param name="sensor_model" value="$(arg sensor_model)" />
        <param name="sensor_height" value="$(arg sensor_height)" />
        <param name="radial_divider_angle" value="$(arg radial_divider_angle)"/>
        <param name="concentric_divider_distance" value="$(arg concentric_divider_distance)"/>
        <param name="local_max_slope" value="$(arg local_max_slope)" />
        <param name="general_max_slope" value="$(arg general_max_slope)" />
        <param name="min_height_threshold" value="$(arg min_height_threshold)" />
        <param name="reclass_distance_threshold" value="$(arg reclass_distance_threshold)" />
        <param name="min_distance" value="$(arg min_distance)"/>
        <param name="clip_height" value="$(arg clip_height)" />
        <param name="minX" value="$(arg minX)" />
        <param name="maxX" value="$(arg maxX)" />
        <param name="minY" value="$(arg minY)" />
        <param name="maxY" value="$(arg maxY)" />
        <param name="debug_pub_all" value="$(arg debug_pub_all)" />
        <param name="debug_map_topic" value="$(arg debug_map_topic)" />
        <!-- <param name="default_cluster_height" value="$(arg default_cluster_height)"/>
        <param name="default_plane_radius" value="$(arg default_plane_radius)"/> -->
    </node>
</launch>
